lib/Conversion/TTIRToTTIRDecomposition/TTIRToTTIRDecomposition.cpp

```diff
@@ -406,9 +407,227 @@ struct ConvolutionToConv2dPattern
   }
 };
 
+struct PoolingToPool2dPattern : public OpConversionPattern<ttir::PoolingOp> {
+public:
+  using OpConversionPattern<ttir::PoolingOp>::OpConversionPattern;
+
+  std::vector<int64_t> getSpatialDims(ttir::PoolingOp op) const {
+    std::vector<int64_t> spatial_dims;
+    for (int64_t i = 0;
+         i < static_cast<int64_t>(op.getWindowDimensions().size()); i++) {
+      if (op.getWindowDimensions()[i] > 1) {
+        spatial_dims.push_back(i);
+      }
+    }
+    return spatial_dims;
+  }
+
+  LogicalResult canDecompose2DPoolingOp(ttir::PoolingOp op) const {
+
+    // Window dimensions must be 4 in length
+    if (op.getWindowDimensions().size() != 4) {
+      return failure();
+    }
+
+    // Window strides must be 4 in length
+    if (op.getWindowStrides().size() != 4) {
+      return failure();
+    }
+
+    // Operand rank(s) must be 4
+    for (Value operand : op.getInputs()) {
+      auto operand_type = mlir::cast<mlir::RankedTensorType>(operand.getType());
+      if (operand_type.getRank() != 4) {
+        return failure();
+      }
+    }
+
+    // Exactly two of the window dimensions must be greater than 1
+    std::vector<int64_t> true_window_dimensions_indices = getSpatialDims(op);
+
+    if (true_window_dimensions_indices.size() != 2) {
+      return failure();
+    }
+
+    // Exactly two of the window strides must be greater than 1
+    std::vector<int64_t> true_window_stride_indices;
+    for (int64_t i = 0; i < static_cast<int64_t>(op.getWindowStrides().size());
+         i++) {
+      if (op.getWindowStrides()[i] > 1) {
+        true_window_stride_indices.push_back(i);
+      }
+    }
+
+    if (true_window_stride_indices.size() != 2) {
+      return failure();
+    }
+
+    // The indices of the true window dimensions and strides must be the same
+    if ((true_window_dimensions_indices[0] != true_window_stride_indices[0] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[1]) &&
+        (true_window_dimensions_indices[0] != true_window_stride_indices[1] ||
+         true_window_dimensions_indices[1] != true_window_stride_indices[0])) {
+      return failure();
+    }
+
+    // Padding must be 8 in length
+    if (op.getPadding().size() != 8) {
```

Comment:
2 padding values per dim = 8 values? Should their indices also match up to window dims&strides?

Reply:
No because of the high and low. We might want to pad the top differently than the bottom even though they exist on the same dimension.

Reply:
Right, 2-sided padding for each dim, just wanted to clarify that. The second question was whether we should have the same test that we do above, but for padding as well, so that we confirm they "see" the same dimensions as spatial.